File: sciurus17_examples_py/sciurus17_examples_py/pick_and_place_right_arm_waist.py
# Copyright 2024 RT Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

from geometry_msgs.msg import Pose
from geometry_msgs.msg import PoseStamped
from moveit.core.robot_state import RobotState
from moveit.planning import MoveItPy
from moveit.planning import PlanRequestParameters
from moveit_msgs.msg import Constraints
from moveit_msgs.msg import JointConstraint
import rclpy
from rclpy.logging import get_logger

from sciurus17_examples_py.utils import plan_and_execute


class PickAndPlace:
    # グリッパの開閉角度
    GRIPPER_OPEN = math.radians(40.0)
    GRIPPER_GRASP = math.radians(20.0)
    GRIPPER_CLOSE = math.radians(0.0)

    def __init__(self):
        # MoveItPyのインスタンスを生成し、planning componentを取得
        self.sciurus17 = MoveItPy(node_name='pick_and_place_right_arm_waist')
        self.logger = get_logger('pick_and_place_right_arm_waist')

        # 右腕・腰・右グリッパ制御用 planning component
        self.r_arm_waist_group = self.sciurus17.get_planning_component(
            'r_arm_waist_group'
        )
        self.r_gripper_group = self.sciurus17.get_planning_component(
            'r_gripper_group'
        )

        # ロボットモデルの取得（ジョイント目標値の設定に使用）
        self.robot_model = self.sciurus17.get_robot_model()

        # プランニングの設定（動作プランナーと速度・加速度スケール）
        self.arm_plan_params = PlanRequestParameters(
            self.sciurus17, 'ompl_rrtc_default'
        )
        self.arm_plan_params.max_velocity_scaling_factor = 0.1  # Set 0.0 ~ 1.0
        self.arm_plan_params.max_acceleration_scaling_factor = (
            0.1  # Set 0.0 ~ 1.0
        )

        self.gripper_plan_params = PlanRequestParameters(
            self.sciurus17, 'ompl_rrtc_default'
        )

    def move_arm_to_pose(self, pose):
        # アームを目標位置・姿勢（Pose）に動かす
        # 座標系はbase_link、目標リンクはr_link7
        self.r_arm_waist_group.set_start_state_to_current_state()
        goal_pose = PoseStamped()
        goal_pose.header.frame_id = 'base_link'
        goal_pose.pose = pose
        self.r_arm_waist_group.set_goal_state(
            pose_stamped_msg=goal_pose,
            pose_link='r_link7',
        )
        plan_and_execute(
            self.sciurus17,
            self.r_arm_waist_group,
            self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

    def control_arm(self, x, y, z):
        # アームを目標位置（x, y, z [m]）に動かす（姿勢は下向き固定）
        pose = Pose()
        pose.position.x = x
        pose.position.y = y
        pose.position.z = z
        pose.orientation.x = 0.707
        pose.orientation.y = 0.0
        pose.orientation.z = 0.0
        pose.orientation.w = 0.707
        self.move_arm_to_pose(pose)

    def move_arm_to_named_pose(self, configuration_name):
        # SRDFに定義された姿勢名でアームを動かす
        self.r_arm_waist_group.set_start_state_to_current_state()
        self.r_arm_waist_group.set_goal_state(
            configuration_name=configuration_name
        )
        plan_and_execute(
            self.sciurus17,
            self.r_arm_waist_group,
            self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

    def move_gripper_angle(self, angle):
        # グリッパを角度[rad]を指定して開閉する
        self.r_gripper_group.set_start_state_to_current_state()
        robot_state = RobotState(self.robot_model)
        robot_state.set_joint_group_positions('r_gripper_group', [angle])
        self.r_gripper_group.set_goal_state(robot_state=robot_state)
        plan_and_execute(
            self.sciurus17,
            self.r_gripper_group,
            self.logger,
            single_plan_parameters=self.gripper_plan_params,
        )

    def set_constraints(self):
        # 腰関節の可動範囲制限を設定する
        constraints = Constraints()
        joint_constraint = JointConstraint()
        joint_constraint.joint_name = 'waist_yaw_joint'
        joint_constraint.position = 0.0
        joint_constraint.tolerance_above = math.radians(45.0)
        joint_constraint.tolerance_below = math.radians(45.0)
        joint_constraint.weight = 1.0
        constraints.joint_constraints.append(joint_constraint)
        self.r_arm_waist_group.set_path_constraints(
            path_constraints=constraints
        )

    def clear_constraints(self):
        # 設定された関節可動制限をクリアする
        self.r_arm_waist_group.set_path_constraints(path_constraints=None)


def main(args=None):
    rclpy.init(args=args)

    controller = PickAndPlace()

    # アプローチ・退避時の高さ
    LIFTING_HEIGHT = 0.25

    # 掴む位置（ピック位置）のXYZ[m]
    PICK_X = 0.25
    PICK_Y = 0.0
    PICK_Z = 0.12

    # 置く位置（プレース位置）のXYZ[m]
    PLACE_X = 0.35
    PLACE_Y = 0.0
    PLACE_Z = 0.12

    # 初期化動作
    controller.move_arm_to_named_pose('r_arm_waist_init_pose')
    controller.move_gripper_angle(controller.GRIPPER_OPEN)
    controller.set_constraints()

    # ピック動作（掴みに行く）
    # 物体の上に腕を伸ばす
    controller.control_arm(PICK_X, PICK_Y, LIFTING_HEIGHT)
    # アプローチ
    controller.control_arm(PICK_X, PICK_Y, PICK_Z)
    # 掴む
    controller.move_gripper_angle(controller.GRIPPER_GRASP)
    # 持ち上げる
    controller.control_arm(PICK_X, PICK_Y, LIFTING_HEIGHT)

    # プレース動作（移動して置く）
    # 移動する
    controller.control_arm(PLACE_X, PLACE_Y, LIFTING_HEIGHT)
    # 下ろす
    controller.control_arm(PLACE_X, PLACE_Y, PLACE_Z)
    # 離す
    controller.move_gripper_angle(controller.GRIPPER_OPEN)
    # 持ち上げる
    controller.control_arm(PLACE_X, PLACE_Y, LIFTING_HEIGHT)

    # 終了動作
    controller.clear_constraints()
    controller.move_arm_to_named_pose('r_arm_waist_init_pose')
    controller.move_gripper_angle(controller.GRIPPER_CLOSE)

    # Finish with error. Related Issue
    # https://github.com/moveit/moveit2/issues/2693
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: sciurus17_examples/src/pick_and_place_right_arm_waist.cpp
// Copyright 2023 RT Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Reference:
// https://github.com/ros-planning/moveit2_tutorials/blob/humble/doc/
// examples/move_group_interface/src/move_group_interface_tutorial.cpp

#include <angles/angles.h>

#include <cmath>
#include <thread>

#include <geometry_msgs/msg/pose.hpp>
#include <moveit/move_group_interface/move_group_interface.hpp>
#include <moveit_msgs/msg/constraints.hpp>
#include <rclcpp/rclcpp.hpp>

#include "pose_presets.hpp"

using MoveGroupInterface = moveit::planning_interface::MoveGroupInterface;

class PickAndPlace
{
public:
  // グリッパの開閉角度
  inline static const double GRIPPER_OPEN = angles::from_degrees(40.0);
  inline static const double GRIPPER_GRASP = angles::from_degrees(20.0);
  inline static const double GRIPPER_CLOSE = 0.0;

  // ノードを受け取り、アーム・グリッパのMoveGroupInterfaceを初期化する
  explicit PickAndPlace(rclcpp::Node::SharedPtr arm_node, rclcpp::Node::SharedPtr gripper_node)
  {
    r_arm_waist_group_ = std::make_shared<MoveGroupInterface>(arm_node, "r_arm_waist_group");
    r_arm_waist_group_->setMaxVelocityScalingFactor(0.1);      // 0.0〜1.0の範囲で設定
    r_arm_waist_group_->setMaxAccelerationScalingFactor(0.1);  // 0.0〜1.0の範囲で設定

    r_gripper_group_ = std::make_shared<MoveGroupInterface>(gripper_node, "r_gripper_group");
  }

  // アームを目標位置・姿勢（Pose）に動かす
  void move_arm_to_pose(const geometry_msgs::msg::Pose & pose)
  {
    r_arm_waist_group_->setPoseTarget(pose);
    r_arm_waist_group_->move();
  }

  // アームを目標位置（x, y, z [m]）に動かす（グリッパは下向き固定）
  void control_arm(const double x, const double y, const double z)
  {
    move_arm_to_pose(pose_presets::right_arm_downward(x, y, z));
  }

  // SRDFに定義された姿勢名でアームを動かす
  void move_arm_to_named_pose(const std::string & name)
  {
    r_arm_waist_group_->setNamedTarget(name);
    r_arm_waist_group_->move();
  }

  // グリッパを角度[rad]を指定して開閉する
  void move_gripper_angle(const double angle)
  {
    auto joint_values = r_gripper_group_->getCurrentJointValues();
    joint_values[0] = angle;
    r_gripper_group_->setJointValueTarget(joint_values);
    r_gripper_group_->move();
  }

  // アームの関節の一部に可動制限を設定する
  void set_constraints()
  {
    moveit_msgs::msg::Constraints constraints;
    constraints.name = "arm_constraints";

    moveit_msgs::msg::JointConstraint joint_constraint;
    joint_constraint.joint_name = "waist_yaw_joint";
    joint_constraint.position = 0.0;
    joint_constraint.tolerance_above = angles::from_degrees(45);
    joint_constraint.tolerance_below = angles::from_degrees(45);
    joint_constraint.weight = 1.0;
    constraints.joint_constraints.push_back(joint_constraint);

    r_arm_waist_group_->setPathConstraints(constraints);
  }

  // 設定された関節可動制限をクリアする
  void clear_constraints() {r_arm_waist_group_->clearPathConstraints();}

private:
  std::shared_ptr<MoveGroupInterface> r_arm_waist_group_;
  std::shared_ptr<MoveGroupInterface> r_gripper_group_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  node_options.automatically_declare_parameters_from_overrides(true);
  auto move_group_arm_node = rclcpp::Node::make_shared("move_group_arm_node", node_options);
  auto move_group_gripper_node = rclcpp::Node::make_shared("move_group_gripper_node", node_options);

  // MoveGroupInterfaceのデッドロックを防ぐため、スピン処理を別スレッドで走らせる
  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(move_group_arm_node);
  executor.add_node(move_group_gripper_node);
  std::thread spin_thread([&executor]() {executor.spin();});

  PickAndPlace controller(move_group_arm_node, move_group_gripper_node);

  // アプローチ・退避時の高さ
  const double LIFTING_HEIGHT = 0.25;

  // 掴む位置（ピック位置）のXYZ[m]
  const double PICK_X = 0.25;
  const double PICK_Y = 0.0;
  const double PICK_Z = 0.12;

  // 置く位置（プレース位置）のXYZ[m]
  const double PLACE_X = 0.35;
  const double PLACE_Y = 0.0;
  const double PLACE_Z = 0.12;

  // 初期化動作
  controller.move_arm_to_named_pose("r_arm_waist_init_pose");
  // 何かを掴んでいた時のために開く
  controller.move_gripper_angle(PickAndPlace::GRIPPER_OPEN);
  controller.set_constraints();

  // ピック動作（掴みに行く）
  // 物体の上に腕を伸ばす
  controller.control_arm(PICK_X, PICK_Y, LIFTING_HEIGHT);
  // アプローチ
  controller.control_arm(PICK_X, PICK_Y, PICK_Z);
  // 掴む
  controller.move_gripper_angle(PickAndPlace::GRIPPER_GRASP);
  // 持ち上げる
  controller.control_arm(PICK_X, PICK_Y, LIFTING_HEIGHT);

  // プレース動作（移動して置く）
  // 移動する
  controller.control_arm(PLACE_X, PLACE_Y, LIFTING_HEIGHT);
  // 下ろす
  controller.control_arm(PLACE_X, PLACE_Y, PLACE_Z);
  // 離す
  controller.move_gripper_angle(PickAndPlace::GRIPPER_OPEN);
  // 持ち上げる
  controller.control_arm(PLACE_X, PLACE_Y, LIFTING_HEIGHT);

  // 終了動作
  controller.clear_constraints();
  controller.move_arm_to_named_pose("r_arm_waist_init_pose");
  controller.move_gripper_angle(PickAndPlace::GRIPPER_CLOSE);

  rclcpp::shutdown();
  spin_thread.join();
  return 0;
}
